faust/tables/recovery.py

```diff
@@ -0,0 +1,236 @@
+
+        for tp in newly_assigned:
+            table = self.tables._changelogs.get(tp.topic)
+            if table is not None:
+                if await table.need_active_standby_for(tp):
+                    if tp in standby_tps:
+                        self.add_standby(tp)
+                    elif tp in assigned_tps:
+                        self.add_active(tp)
+                    else:
+                        pass   # belongs to agent?
+
+        print('ACTIVE TPS: %r' % (self.active_tps,))
+        print('STANDBY TPS: %r' % (self.standby_tps,))
+        print('ALL TPS: %r' % (self.tps,))
+
+        # Pause all table changelog partitions
+        await consumer.pause_partitions(tps)
+        await self._build_highwaters(consumer)
+        await self._build_offsets(consumer)
+
+        # Seek to new offsets
+        for tp in tps:
+            offset = offsets[tp]
+            if offset >= 0:
+                # FIXME Remove check when fixed offset-1 discrepancy
+                await consumer.seek(tp, offset)
+                assert await consumer.position(tp) == offset
+
+        self.log.info('Restoring state from changelog topics...')
+
+        # Resume partitions and start fetching.
+        await consumer.resume_partitions(tps)
+        await self.app._fetcher.start()
+
+        await self.wait_for_actives()
+
+        self.log.info('Done reading from changelog topics')
+        # Pause all our topic partitions,
+        # to make sure we don't fetch any more records from them.
+        assignment = self.app.consumer.assignment()
+        await consumer.pause_partitions(
+            {tp for tp in tps if tp in assignment})
+        await self.tables.on_recovery_completed()
+
+    async def wait_for_actives(self):
+        print('SHOULD WAIT FOR ACTIVES')
+
+    @Service.task
+    async def _slurp_changelogs(self):
+        changelog_queue = self.tables.changelog_queue
+        tp_to_table = self.tp_to_table
+        offsets = self.offsets
+
+        buffers: MutableMapping[CollectionT, List[EventT]] = defaultdict(list)
+
+        buffer_sizes: MutableMapping[TP, int] = {}
+
+        while not self.should_stop:
+            event: EventT = await changelog_queue.get()
+            message = event.message
+            tp = message.tp
+            offset = message.offset
+            seen_offset = offsets.get(tp, -1)
+            if offset > seen_offset:
+                offsets[tp] = offset
+
+                table = tp_to_table[tp]
+                bufsize = buffer_sizes.get(tp)
+                if bufsize is None:
+                    if tp in self.active_tps:
+                        bufsize = buffer_sizes[tp] = table.active_buffer_size
+                    else:
+                        bufsize = buffer_sizes[tp] = table.standby_buffer_size
+                buf = buffers[table]
+                buf.append(event)
+                await table.on_changelog_event(event)
+                if len(buf) >= bufsize:
```

Comment:
@ask Could we get into the case in which `event: EventT = await changelog_queue.get()` has no more events and `len(buf) < bufsize` and we miss the last `apply_changelog_batch`?

Reply:
Yes! Good catch! we need to take "remaining" into account here, so if we are expecting just 3 more events, we should flushing the buffer.